contracts/SnappAuction.sol

```diff
@@ -0,0 +1,81 @@
+pragma solidity ^0.5.0;
+
+import "./SnappBase.sol";
+
+
+contract SnappAuction is SnappBase {
+  
+    uint16 public constant AUCTION_BATCH_SIZE = 1000;
+
+    uint public auctionIndex;
+    mapping (uint => PendingFlux) public auctions;
+
+    event SellOrder(
+        uint auctionId, 
+        uint16 slotIndex,
+        uint16 accountId, 
+        uint8 buyToken, 
+        uint8 sellToken, 
+        uint128 buyAmount,
+        uint128 sellAmount
+    );
+
+    constructor () public {
+        auctions[auctionIndex].creationBlock = block.number;
+    }
+
+    function placeSellOrder(
+        uint8 buyToken,
+        uint8 sellToken,
+        uint128 minBuyAmount,
+        uint128 maxSellAmount
+    ) public onlyRegistered() {
+        require(buyToken != sellToken, "Buy and Sell tokens must differ!");
+
+        // Must have 0 < tokenId < MAX_TOKENS anyway, so might as well ensure registered.
+        require(tokenIdToAddressMap[buyToken] != address(0), "Buy token is not registered");
+        require(tokenIdToAddressMap[sellToken] != address(0), "Sell token is not registered");
+
+        // Restrict buy and sell amount to occupy less than 100 bits.
+        require(minBuyAmount < 0x10000000000000000000000000, "Buy amount too large!");
+        require(maxSellAmount < 0x10000000000000000000000000, "Sell amount too large!");
+
+        if (
+            auctions[auctionIndex].size == AUCTION_BATCH_SIZE || 
+            block.number > auctions[auctionIndex].creationBlock + 20
```

Comment:
I'm a bit concerned about this condition. This could lead to a devil spiral if our processing time takes longer than 20 blocks. Then we would create new batches faster then we process them leading to an evergrowing backlog.

I wonder if it would make sense to somehow limit new batch creation to a maximum of previous unapplied batches.

Reply:
What do you mean by processing time takes longer? Isn't the processing time determined by the smart contract on how long it waits for "highest pids" for traders welfare? We just need to make sure that these two numbers, the 20 from here and the submission time for traders welfare are fitting together.

But for the escalation and stopping of chain, such a condition might make for sure a lot of sense

Reply:
I can get behind this. I will have to go back to the drawing board here though and think about a new approach.

Reply:
I'd suggest checking that for auctionIndex > 2, `auctions[auctionIndex -2].shaHash != 0`, which would imply that the auction has been applied.

Reply:
I don't see how non-zero shaHash implies auction has been applied.

Reply:
Why not simply reject all new orders until the previous batch has been applied? Or maybe allow for 3 batches to be collected before rejecting new orders.

I believe the condition for a batch having been applied is 

```
auctions[auctionIndex - 2].appliedAccountStateIndex != 0
```

(Now I see what you were referring to with the `.shaHash` @fleupold ).

Reply:
This should cover it

```js
        if (auctionIndex > 1) {
            require(auctions[auctionIndex - 2].appliedAccountStateIndex != 0, "To many pending auctions");
        }
```

Reply:
Yes that's what I mean. Great!